hal/ia32/string.c

```diff
@@ -0,0 +1,148 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * HAL basic routines
+ *
+ * Copyright 2012, 2021 Phoenix Systems
+ * Copyright 2001, 2005-2006 Pawel Pisarczyk
+ * Author: Pawel Pisarczyk, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include "string.h"
+
+
+void *hal_memcpy(void *dst, const void *src, size_t n)
+{
+	__asm__ volatile(
+		"cld; "
+		"rep; movsl; "
+		"movl %%edx, %%ecx; "
+		"rep; movsb; "
+	:: "c" (n >> 2), "d" (n & 3), "S" (src), "D" (dst)
+	: "memory", "cc");
+
+	return dst;
+}
+
+
+void *hal_memset(void *dst, int v, size_t n)
+{
+	__asm__ volatile(
+		"cld; "
+		"movl %%eax, %%ebx; "
+		"shll $0x8, %%ebx; "
+		"orl %%ebx, %%eax; "
+		"movl %%eax, %%ebx; "
+		"shll $0x10, %%ebx; "
+		"orl %%ebx, %%eax; "
+		"rep; stosl; "
+		"movl %%edx, %%ecx; "
+		"rep; stosb; "
+	:: "a" (v & 0xff), "c" (n >> 2), "d" (n & 3), "D" (dst)
+	: "ebx", "memory", "cc");
```

Comment:
This PR is marked as a draft but necessary fix.
```suggestion
	void *d = *dst;
	
	__asm__ volatile(
		"cld; "
		"movl %%eax, %%ebx; "
		"shll $0x8, %%ebx; "
		"orl %%ebx, %%eax; "
		"movl %%eax, %%ebx; "
		"shll $0x10, %%ebx; "
		"orl %%ebx, %%eax; "
		"rep; stosl; "
		"movl %%edx, %%ecx; "
		"rep; stosb; "
	: "+D" (d)
	: "a" (v & 0xff), "c" (n >> 2), "d" (n & 3)
	: "ebx", "memory", "cc");
```
In the current version, you return `dst + n`, not `dst`.
Look at the end of the code disassembly:
```
  26:   f3 ab                   rep stos %eax,%es:(%edi)
  28:   89 d1                   mov    %edx,%ecx
  2a:   f3 aa                   rep stos %al,%es:(%edi)
  2c:   89 f8                   mov    %edi,%eax
  2e:   5b                      pop    %ebx
  2f:   5f                      pop    %edi
  30:   c3                      ret    
```
For safety, you should also use an extra variable in the `hal_memcpy()` function and pass it to the `edi` register like here.

Reply:
I totally agree - thanks for finding this bug. It's fixed now in both `hal_mecpy()` and `hal_memset()`.